Assets/MixedRealityToolkit/_Core/Utilities/ClipPlane.cs

```diff
@@ -62,95 +76,114 @@ private void OnDrawGizmosSelected()
                 }
             }
+        }
 
+        private void Initialize()
+        {
+            clipPlaneID = Shader.PropertyToID("_ClipPlane");
             materialPropertyBlock = new MaterialPropertyBlock();
         }
 
         private void UpdatePlanePosition()
         {
-            if (renderers == null)
+            if (Renderers == null)
             {
                 return;
             }
 
             Vector3 up = transform.up;
             Vector4 plane = new Vector4(up.x, up.y, up.z, Vector3.Dot(up, transform.position));
 
-            foreach (Renderer renderer in renderers)
+            foreach (Renderer renderer in Renderers)
             {
-                if (renderer == null)
+                if (!renderer)
                 {
                     continue;
                 }
 
                 renderer.GetPropertyBlock(materialPropertyBlock);
-                materialPropertyBlock.SetVector(clipPlaneId, plane);
+                materialPropertyBlock.SetVector(clipPlaneID, plane);
                 renderer.SetPropertyBlock(materialPropertyBlock);
             }
         }
 
-        private void ToggleClippingPlane(bool isClippingPlaneOn)
+        private void ToggleClippingPlanes(bool clippingPlaneOn)
         {
-            if (materials == null)
+            if (Renderers == null)
             {
                 return;
             }
 
-            foreach (Material material in materials)
+            foreach (Renderer renderer in Renderers)
             {
-                if (material == null)
+                if (!renderer)
                 {
                     continue;
                 }
 
-                const string clippingPlaneKeyword = "_CLIPPING_PLANE";
+                Material material = GetMaterial(renderer);
 
-                if (isClippingPlaneOn)
+                if (material)
                 {
-                    if (!material.IsKeywordEnabled(clippingPlaneKeyword))
+                    // Cache the initial keyword state of the material.
+                    if (!modifiedMaterials.ContainsKey(material))
                     {
-                        material.EnableKeyword(clippingPlaneKeyword);
-                    }
-                }
-                else
-                {
-                    if (material.IsKeywordEnabled(clippingPlaneKeyword))
-                    {
-                        material.DisableKeyword(clippingPlaneKeyword);
+                        modifiedMaterials[material] = material.IsKeywordEnabled(clippingPlaneKeyword);
                     }
+
+                    ToggleClippingPlaneKeyword(material, clippingPlaneOn);
                 }
             }
         }
+
+        private Material GetMaterial(Renderer renderer)
+        {
+            if (Application.isEditor && !Application.isPlaying)
+            {
+                return renderer.sharedMaterial;
+            }
+            else
+            {
+                return renderer.material;
```

Comment:
this creates a new copy of that material. Are we disposing that material correctly to avoid the memory leak?

Reply:
Ah, no we aren't. I'll loop though modifiedMaterials OnDestroy and clean them up there.